Invalid cast error when trying to cast a closure to a function pointer should mention captured variables

### Code

```Rust
fn main() {
    let captured = "x";
    let closure = || { captured.to_string() };
    let fn_ptr = closure as fn() -> String;
    let _ = fn_ptr();
}
```


### Current output

```Shell
error[E0605]: non-primitive cast: `{closure@src/main.rs:3:19: 3:21}` as `fn() -> String`
 --> src/main.rs:4:18
  |
4 |     let fn_ptr = closure as fn() -> String;
  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^ invalid cast

For more information about this error, try `rustc --explain E0605`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```


### Desired output

```Shell
error[E0605]: non-primitive cast: `{closure@src/main.rs:3:19: 3:21}` as `fn() -> String`
 --> src/main.rs:4:18
  |
4 |     let fn_ptr = closure as fn() -> String;
  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^ invalid cast
help: this closure cannot be cast to a function pointer because it captures a variable
  |
2 |     let captured = "x";
  |         ^^^^^^^- captured outer variable
3 |     let closure = || { captured.to_string() }; 
  |                   -- captured by this Fn closure

For more information about this error, try `rustc --explain E0605`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```


### Rationale and extra context

I forgot a closure was capturing a variable while I was refactoring and setting up an array of function pointers. I had to double check the [reference](https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions) to figure out what the error was referring to.

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: x86_64-unknown-linux-gnu
release: 1.79.0
LLVM version: 18.1.7
```


### Anything else?

_No response_

--- a/compiler/rustc_hir_typeck/src/cast.rs
+++ b/compiler/rustc_hir_typeck/src/cast.rs
@@ -495,6 +495,7 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                     err.span_label(self.span, "invalid cast");
                 }
 
+                fcx.suggest_no_capture_closure(&mut err, self.cast_ty, self.expr_ty);
                 self.try_suggest_collection_to_bool(fcx, &mut err);
 
                 err.emit();

--- a/tests/ui/closures/closure-no-fn-3.stderr
+++ b/tests/ui/closures/closure-no-fn-3.stderr
@@ -3,6 +3,12 @@ error[E0605]: non-primitive cast: `{closure@$DIR/closure-no-fn-3.rs:6:28: 6:30}`
    |
 LL |     let baz: fn() -> u8 = (|| { b }) as fn() -> u8;
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^ invalid cast
+   |
+note: closures can only be coerced to `fn` types if they do not capture any variables
+  --> $DIR/closure-no-fn-3.rs:6:33
+   |
+LL |     let baz: fn() -> u8 = (|| { b }) as fn() -> u8;
+   |                                 ^ `b` captured here
 
 error: aborting due to 1 previous error
 

